_posts/2021-01-08-consumer-tuning.md

```diff
@@ -0,0 +1,292 @@
+
+* **`fetch.max.bytes`**
+Sets a maximum limit in bytes on the amount of data fetched from the broker at one time
+
+* **`max.partition.fetch.bytes`**
+Sets a maximum limit in bytes on how much data is returned for each partition, which must always be larger than the number of bytes set in the broker or topic configuration for `max.message.bytes`.
+
+By increasing the values of these two properties, and allowing more data in each request, latency might be improved as there are fewer fetch requests.
+
+```properties
+# ...
+fetch.max.bytes=52428800
+max.partition.fetch.bytes=1048576
+# ...
+```
+
+> Be careful with memory usage. The maximum amount of memory a client can consume is calculated approximately as:
+>
+> <pre><i>NUMBER-OF-BROKERS</i> * fetch.max.bytes and <i>NUMBER-OF-PARTITIONS</i> * max.partition.fetch.bytes</pre>
+
+
+### Avoiding data loss or duplication when committing offsets
+
+For applications that require durable message delivery, you can increase the level of control over consumers when committing offsets to minimize the risk of data being lost or duplicated.
+
+Kafka's _auto-commit mechanism_ is pretty convenient (and sometimes suitable). When enabled, consumers commit the offsets of messages automatically. But convenience, as always, has a price. By allowing your consumer to commit offsets, you are introducing a risk of data loss and duplication. Why? Quite easily as it happens. You see, the auto-commit can _only_ avoid data loss if all messages are processed before the next poll of the broker or the consumer closes. Using the auto-commit, the consumer will commit offsets received from the broker at 5000ms intervals. In that time, the inherent constraints of auto-committing might lead to:
+
+* _Data loss_ when the consumer has fetched and handled a number of messages, but the system crashes with processed messages still in the consumer buffer
+* _Data duplication_ on another consumer instance, after rebalancing, when the system crashes whilst processing the messages, but before performing the auto-commit
+
+If this potential situation leaves you slightly concerned, what can you do about it? First of all, you can use the `auto.commit.interval.ms` property to decrease those worrying intervals between commits. But this will not completely eliminate the chance that messages are lost or duplicated.
+
+What else? Alternatively, you can turn off auto-committing by setting `enable.auto.commit` to `false`. You then assume responsibility for how your consumer application handles commits correctly. This is an important decision. Periodic automatic offset commits do mean it's something you don't need to worry about. And as long as all message processing is done before the next poll, all processed offsets will be committed. But a higher level of control might be preferable if data loss or data duplication is to be avoided.   
+
+```properties
+# ...
+enable.auto.commit=false
+auto.commit.interval.ms=1000
+# ...
```

Comment:
I would move the `auto.commit.interval.ms=1000` example at the end of the paragraph where we mention it and leave the auto commit disabling here, right after the paragraph mentioning it.
Having this example with both changes could be misleading (I am disabling but reducing the commit interval??)

Reply:
Good idea. I've separated.